doc/assumeutxo.md

```diff
@@ -0,0 +1,138 @@
+# assumeutxo
+
+Assumeutxo is a feature that allows fast bootstrapping of a validating bitcoind
+instance with a very similar security model to assumevalid.
+
+The RPC commands `dumptxoutset` and `loadtxoutset` are used to respectively generate
+and load UTXO snapshots. The utility script `./contrib/devtools/utxo_snapshot.sh` may
+be of use.
+
+## General background
+
+- [assumeutxo proposal](https://github.com/jamesob/assumeutxo-docs/tree/2019-04-proposal/proposal)
+- [Github issue](https://github.com/bitcoin/bitcoin/issues/15605)
+- [draft PR](https://github.com/bitcoin/bitcoin/pull/15606)
+
+## Design notes
+
+- A new block index `nStatus` flag is introduced, `BLOCK_ASSUMED_VALID`, to mark block
```

Comment:
I am wondering who is the audience of this document?

If it is a developer, it might be better to put the documentation in the source code. Otherwise it will be hard for developers to find it and it will more easily get out of date.

If the target audience is a user, I am wondering why internal implementation details are mentioned and why they are relevant to the user.

If the target audience is a reviewer of your pull request, I am wondering why this needs a doc at all in the source tree. Wouldn't a pull request description or so be a better place?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/23154#discussion_r741673855

> I am wondering who is the audience of this document? [...]

I think the concerns here go a little beyond scope of this PR. Some of the files in the doc/ folder are clearly targeted at developers and not helpful to users like https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md. Some files like https://github.com/bitcoin/bitcoin/blob/master/doc/JSON-RPC-interface.md and the release notes are mostly useful for end users, and definitely intended to be read by them. I think ideally someone would go through the doc folder, keeping the files that are intended to be read by users but moving the other ones that are intended for developers/testers/contributors somewhere else like `doc/devel`,`doc/internal`, `doc/contrib`, `doc/design`.

This document is a basically a design document (other examples: [golang](https://go.googlesource.com/proposal/+/master/design), [chromium](https://www.chromium.org/developers/design-documents)) intended to by read by developers. It should be useful not just to people reviewing the PR, but also anybody going forward who want to understand how this feature which touches different parts of the source code is implemented and what some of design considerations are.

At the top of this file, there is a small amount of information that could be useful to users, but it is brief and basically just linking to RPCs and a script. If bitcoin had a user manual like [many other projects](https://www.gnu.org/manual/manual.en.html) do, this section could be copied there. But bitcoin doesn't have a user manual so, 🤷

> If it is a developer, it might be better to put the documentation in the source code. Otherwise it will be hard for developers to find it and it will more easily get out of date.

I think this could be addressed by organizing the doc directory a little better and linking to the document from the source code so it is easier to find.

> If the target audience is a user, I am wondering why internal implementation details are mentioned and why they are relevant to the user.

Users are not target audience here, AFAICT.

> If the target audience is a reviewer of your pull request, I am wondering why this needs a doc at all in the source tree. Wouldn't a pull request description or so be a better place?

I think this is a false dichotomy. If I am reviewing a pull request that includes end-user release notes, those release notes are usually more helpful to me for understanding behavior of the PR than any of the weedsy implementation details in the PR description or other parts of the diff.

PR descriptions do tend to be better places than design documents for certain information, like describing details or drawbacks of previous code, but design documents are useful _both_ for current developers reviewing changes, and for future developers want to understand how a feature works. It is not an either/or.